### ISSUE TITLE
hadleyverse: cannot install packages as 'lib' is unspecified and site libs are not writeable

### ISSUE THREAD

**Issue by james-woods** [rocker-org/rocker#207]
Using the latest tag I was trying to install packages from a vanilla hadleyverse image 

```
> install.packages('rmongodb')
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
```
- do I have to specify this manually now? In a previous version it worked out of the box.

**Comment by eddelbuettel** [comment]
That is a purely informational message you are misreading.   

Witness:

```
~> cd /usr/local/lib/R/site-library/
/usr/local/lib/R/site-library> ls -1 | wc -l
383
/usr/local/lib/R/site-library> rm -rf pkgKitten/    # random pick
/usr/local/lib/R/site-library> Rscript -e 'install.packages("pkgKitten")'
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://cran.rstudio.com/src/contrib/pkgKitten_0.1.4.tar.gz'
Content type 'application/x-gzip' length 7074 bytes
==================================================
downloaded 7074 bytes

* installing *source* package ‘pkgKitten’ ...
** package ‘pkgKitten’ successfully unpacked and MD5 sums checked
** R
** demo
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
* DONE (pkgKitten)

The downloaded source packages are in
        ‘/tmp/RtmpE4jq2C/downloaded_packages’
/usr/local/lib/R/site-library> ls pkgKitten/
demo  DESCRIPTION  help  html  INDEX  Meta  NAMESPACE  NEWS.Rd  R  replacements
/usr/local/lib/R/site-library> 
```

**Comment by james-woods** [comment]
You are right, that is true, for the standard user. When mounting external directories on the host and mapping the user id this does not work. 

docker-compose.yml:
```
version: '2'
services:
    rserver:
        image: rocker/hadleyverse:latest
        ports:
          - "0.0.0.0:8787:8787"
        environment:
          - USER=someuser
          - USERID=1111
          - ROOT=TRUE
        volumes:
          - /data/rserver:/home/someuser/data
          - /home/someuser/data_rcode:/home/someuser/rcode
```

Output of `install.packages('pkgKitten')`

```
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
Warning in install.packages :
  'lib = "/usr/local/lib/R/site-library"' is not writable
Would you like to use a personal library instead?  (y/n) y
trying URL 'https://cran.rstudio.com/src/contrib/pkgKitten_0.1.4.tar.gz'
Content type 'unknown' length 7074 bytes
==================================================
downloaded 7074 bytes

Error: ERROR: no permission to install to directory ‘/usr/local/lib/R/site-library’
Warning in install.packages :
  installation of package ‘pkgKitten’ had non-zero exit status

The downloaded source packages are in
	‘/tmp/Rtmp1sIVGk/downloaded_packages’
```

**Comment by eddelbuettel** [comment]
Methinks that is your use case.  Ensure you have correct permission across network boundaries and user ids.

**Comment by james-woods** [comment]
Even if I just specify the $USERID as documented [here](https://github.com/rocker-org/rocker/wiki/Sharing-files-with-host-machine#linux) it breaks in the same way.

How can I ensure correct permission across network boundaries and user ids inside of the container when I supply this parameter?

**Comment by eddelbuettel** [comment]
That is beyond the specs of this project.  You are having a generic Docker across nfs and whatnot issue. 

Sorry.

**Comment by james-woods** [comment]
There is no network involved. There is no NFS, just plain Docker AUFS backed by extfs. I just specified another UID as it was stated in the documentation of this project. I didn't even have to mount any folder. My configuration looks like this:

```
version: '2'
services:
    rserver:
        image: rocker/hadleyverse:latest
        ports:
          - "0.0.0.0:8787:8787"
        environment:
          - USERID=1111
```

This functionality is [implemented in the userconf.sh](https://github.com/rocker-org/rstudio-daily/commits/4108b889997cc1af3e9c09e989303a3ff9012d8b/userconf.sh) and documented [here in the offical documentation](https://github.com/rocker-org/rocker/wiki/Sharing-files-with-host-machine#linux).
Just setting the env variable USERID to something else than 1000 is enough in my case to break the functionality of `install.packages` because of permission issues. In my opinion it would be helpful if this is noted in the documentation.

**Comment by eddelbuettel** [comment]
Ok, maybe @cboettig will be able to help you on the `userconf.sh` side of things.  I was piping in for the parts I (also) cover as the Debian maintainer of the package where we use that `/usr/local/lib/R/` location generally just fine (but with my user hat on I know the group-write issues in a multi-user env.).

**Comment by cboettig** [comment]
@james-woods sorry for the difficulties.  can you check with the default USER name (rstudio) an default USERID first?

**Comment by james-woods** [comment]
This works:
```
version: '2'
services:
    rserver:
        image: rocker/hadleyverse:latest
        ports:
          - "0.0.0.0:8787:8787"
        environment:
          - USER=rstudio
          - PASSWORD=rstudio
          - USERID=1000
```

this does not:

```
version: '2'
services:
    rserver:
        image: rocker/hadleyverse:latest
        ports:
          - "0.0.0.0:8787:8787"
        environment:
          - USER=someuser
          - PASSWORD=rstudio
          - USERID=1111
```

Reason is most likely, that someuser is not part of `staff` which seems to be the owner group of `/usr/local/lib/R/site-library`, currently digging deeper

**Comment by james-woods** [comment]
Running `usermod -a -G staff someuser` inside the container fixes the issue, now someuser is able to write into `/usr/local/lib/R/site-library` and can install packages. I can create a pull request with a fix if you want.

**Comment by eddelbuettel** [comment]
Well, isn't that a move back to square one -- and really your local isse as you and only you have the need for `someuser` ?

We work as shipped with the default user.

**Comment by james-woods** [comment]
Ok, then I would suggest editing [this](https://github.com/rocker-org/rocker/wiki/Sharing-files-with-host-machine#linux) in your Wiki, as it does break the ability to install packages without adding the user to the right group.

It seems like this paragraph is targeted at all users that have the need to share files with the host and make the permissions match the right $UID/$USER.

**Comment by cboettig** [comment]
@james-woods thanks, thought that was the likely issue.  Looks like I forgot to include the `staff` group when creating a new user.  Fixed now.

Note that I now don't think there's really much of a use case for using a non-default user; other than the aesthetic perhaps.  When working with local volumes all that matters is just matching the USERID that matters, not the user name, to ensure permissions are unaltered.  Perhaps the wiki should say this too rather than describing the `USER` config.  

Edits to the Wiki are welcome! It is, after all, a wiki.

**Comment by james-woods** [comment]
@cboettig  thanks for fixing! 👍

**Comment by james-woods** [comment]
@cboettig just to be clear, in the current version, this does also _not_ work for me, so the usermod command will fix this as well

```
version: '2'
services:
    rserver:
        image: rocker/hadleyverse:latest
        ports:
          - "0.0.0.0:8787:8787"
        environment:
          - USER=rstudio
          - PASSWORD=rstudio
          - USERID=1111
```

**Comment by james-woods** [comment]
I think it's due to a failing command.

```
rserver_1  | [cont-init.d] conf: executing...
rserver_1  | deleting user rstudio
rserver_1  | creating new rstudio with UID 1111
rserver_1  | useradd: warning: the home directory already exists.
rserver_1  | Not copying any file from skel directory into it.
rserver_1  | mkdir: cannot create directory ‘/home/rstudio’: File exists
rserver_1  | [cont-init.d] conf: exited 0.
```

https://github.com/rocker-org/rstudio-daily/blob/4108b889997cc1af3e9c09e989303a3ff9012d8b/userconf.sh#L13

This could be the issue, the user is never added to the right group then if the userid is not equal 1000.

**Comment by james-woods** [comment]
@cboettig  I think when you add `usermod -a -G staff $USER` after line 13, you will fix the issue for other $UIDs. There is no need for any other fix, as renaming a user makes him stay in his current groups, so he also stays in `staff`.

I just confirmed, that just changing the username and keeping the UID constant does *not* break the thing. Just changing the $UID does.

This works fine:
```
version: '2'
services:
    rserver:
        image: rocker/hadleyverse:latest
        ports:
          - "0.0.0.0:8787:8787"
        environment:
          - USER=someuser
          - PASSWORD=rstudio
          - USERID=1000
```
Patching after line 13 should do the trick.

**Comment by cboettig** [comment]
@james-woods, Right, thanks.  To change the UID we delete the old default `rstudio` user and create a new one; and that's the point where I forgot to also make the recreated user a member of `staff`.  Like you say, when we just have a custom user with default pid, we rename the user rather than deleting it so the group comes with it.  Sorry I hadn't tested the custom PID thoroughly and thanks for the bug report.  pushed changes now.

**Comment by sankarachari** [comment]
Can any one guide me how to fix the issue of : **install.packages("nnet")**
Installing package into ‘C:/Users/Acer/Documents/R/win-library/3.4’
(as ‘lib’ is unspecified)
Warning: unable to access index for repository https://ftp.iitm.ac.in/cran/src/contrib:
  cannot open URL 'https://ftp.iitm.ac.in/cran/src/contrib/PACKAGES'

Warning message:
package ‘mlogit’ is not available (for R version 3.4.2)

**Comment by eddelbuettel** [comment]
Try `man chmod`, or in your case 'how do I install as admin'. 

I also answered that a few times over the years, the following link has a few relevant ones:
https://stackoverflow.com/search?q=is%3Aanswer+user%3Ame+%2Fusr%2Flocal%2Flib%2FR%2Fsite-library+install.packages

In essence you try to write where you cannot.  It is a fight between you and your windows computer.  As it is on Windows, whereas this project is for Docker (== Unix) you are also abusing this help facility.  

Not cool.
